lib/fluent/plugin/out_azure-storage-append-blob.rb

```diff
@@ -3,122 +3,157 @@
+      config_param :auto_create_container, :bool, default: true
+      config_param :format, :string, default: DEFAULT_FORMAT_TYPE
+      config_param :time_slice_format, :string, default: '%Y%m%d'
+      config_param :localtime, :bool, default: false
+
       config_section :format do
         config_set_default :@type, DEFAULT_FORMAT_TYPE
       end
-  
+
       config_section :buffer do
         config_set_default :chunk_keys, ['time']
         config_set_default :timekey, (60 * 60 * 24)
       end
-  
+
       attr_reader :bs
-  
+
       def configure(conf)
         super
-  
+
         @formatter = formatter_create
-  
-        if @localtime
-          @path_slicer = Proc.new {|path|
-            Time.now.strftime(path)
-          }
-        else
-          @path_slicer = Proc.new {|path|
-            Time.now.utc.strftime(path)
-          }
-        end
-  
-        if @azure_container.nil?
-          raise ConfigError, 'azure_container is needed'
-        end
+
+        @path_slicer = if @localtime
+                         proc do |path|
+                           Time.now.strftime(path)
+                         end
+                       else
+                         proc do |path|
+                           Time.now.utc.strftime(path)
+                         end
+                       end
+
+        raise ConfigError, 'azure_storage_account needs to be specified' if @azure_storage_account.nil?
+
+        raise ConfigError, 'azure_container needs to be specified' if @azure_container.nil?
 
         if @azure_storage_access_key.nil? && @azure_storage_sas_token.nil?
-          raise ConfigError, "either 'azure_storage_access_key' or 'azure_storage_sas_token' parameter must be provided"
+          log.info 'Using MSI since neither azure_storage_access_key nor azure_storage_sas_token was provided.'
+          @use_msi = true
         end
       end
-  
+
       def multi_workers_ready?
         true
       end
-  
+
+      def get_access_token
+        access_key_request = Faraday.new('http://169.254.169.254/metadata/identity/oauth2/token?' \
+                                         'api-version=2019-08-15' \
+                                         '&resource=https://storage.azure.com/',
+                                         headers: { 'Metadata' => 'true' })
+                                    .get
+                                    .body
+        JSON.parse(access_key_request)['access_token']
+      end
+
       def start
         super
-  
-        @bs_params = {storage_account_name: @azure_storage_account}
+        if @use_msi
+          token_credential = Azure::Storage::Common::Core::TokenCredential.new get_access_token
+          token_signer = Azure::Storage::Common::Core::Auth::TokenSigner.new token_credential
+          @bs = Azure::Storage::Blob::BlobService.new(storage_account_name: @azure_storage_account, signer: token_signer)
+
+          # Refresh interval is 50 minutes
+          refresh_interval = 50 * 60
```

Comment:
refresh interval should be configurable but probably defaulted 60

Reply:
This is taken straight from the example provided in the azure storage ruby sdk, see [here](https://github.com/Azure/azure-storage-ruby/tree/master/blob#access-token). I can make it configurable but the tokens I have encountered so far are all valid for 23h59.